ports/rp2/mphalport.c

```diff
@@ -273,21 +287,38 @@ uint32_t storage_read_blocks(uint8_t *dest, uint32_t block_num, uint32_t num_blo
     panic_unsupported();
 }
 
-static int64_t soft_timer_callback(alarm_id_t id, void *user_data) {
-    soft_timer_alarm_id = 0;
-    pendsv_schedule_dispatch(PENDSV_DISPATCH_SOFT_TIMER, soft_timer_handler);
-    return 0; // don't reschedule this alarm
-}
-
 uint32_t soft_timer_get_ms(void) {
     return mp_hal_ticks_ms();
 }
 
 void soft_timer_schedule_at_ms(uint32_t ticks_ms) {
-    if (soft_timer_alarm_id != 0) {
-        cancel_alarm(soft_timer_alarm_id);
-    }
     int32_t ms = soft_timer_ticks_diff(ticks_ms, mp_hal_ticks_ms());
     ms = MAX(0, ms);
-    soft_timer_alarm_id = add_alarm_in_ms(ms, soft_timer_callback, NULL, true);
+    if (hardware_alarm_set_target(MICROPY_HW_SOFT_TIMER_ALARM_NUM, delayed_by_ms(get_absolute_time(), ms))) {
+        // "missed" hardware alarm target
+        hardware_alarm_force_irq(MICROPY_HW_SOFT_TIMER_ALARM_NUM);
+    }
+}
+
+void soft_timer_init(void) {
+    hardware_alarm_claim(MICROPY_HW_SOFT_TIMER_ALARM_NUM);
+    hardware_alarm_set_callback(MICROPY_HW_SOFT_TIMER_ALARM_NUM, (void *)soft_timer_handler);
+    // soft timer IRQ handler has to run at PendSV priority. adjusting priority here so we don't
+    // need to trigger PendSV the timer IRQ.
+    NVIC_SetPriority(TIMER_IRQ_0_IRQn + MICROPY_HW_SOFT_TIMER_ALARM_NUM, PICO_LOWEST_IRQ_PRIORITY);
```

Comment:
A weird thing here: the *real* PendSV IRQ priority depends on whether the network support is compiled in. It's set here in modnetwork.c:
https://github.com/micropython/micropython/blob/master/ports/rp2/mpnetworkport.c#L69

Practically I don't think this matters too much as (AFAIK) it's not used for anything else on rp2 otherwise, but it's a bit odd so I wanted to point it out (which is why I've done it this way here as well.)

Reply:
This is fine.  But the comment on lines 306-307 needs rewriting, I don't understand it!

Reply:
@dpgeorge Right, it had at least one typo in it to boot. PTAL!